### PR TITLE
feat: Add `count_of_trailing_ciphers_in_factorial_n`

### PR DESCRIPTION
**Pull request by Swastyy**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1543"></a>

### CODE REVIEW

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        count += static_cast<uint64_t>(n) / i;
+    }
+
+    return count;
+}
+}  // namespace numberOfCiphersInFactorialN
+}  // namespace bit_manipulation
+
+/**
+ * @brief Test implementations
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:45]:
```suggestion
 * @brief Self-test implementations
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
+/**
+ * @file
+ * @brief Program to count number of ciphers in `n!`(https://www.tutorialspoint.com/count-trailing-zeros-in-factorial-of-a-number-in-cplusplus).
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:3]:
```suggestion
 * @brief [Count the number of ciphers](https://www.tutorialspoint.com/count-trailing-zeros-in-factorial-of-a-number-in-cplusplus) in `n!` implementation
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+ * @details
+ * We are given a number. The algorithm finds all
+ * the zeroes at the end of factorial of that number.
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:6]:
```suggestion
 * Given an integer number as input. The goal is to find the number of trailing zeroes in the factorial calculated for
 * that number. A factorial of a number N is a product of all numbers in the range [1, N].

 * We know that we get a trailing zero only if the number is multiple of 10 or has a factor pair (2,5). In all factorials of
 * any number greater than 5, we have many 2s more than 5s in the prime factorization of that number. Dividing a
 * number by powers of 5 will give us the count of 5s in its factors. So, the number of 5s will tell us the number of trailing zeroes.
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+#include <iostream>  /// for IO operations
+
+/**
+ * @namespace bit_manipulation
+ * @brief Bit manipulation algorithms
+ */
+namespace bit_manipulation {
+/**
+ * @namespace numberOfCiphersInFactorialN
+ * @brief Program to count number of ciphers in `n!`(https://www.tutorialspoint.com/count-trailing-zeros-in-factorial-of-a-number-in-cplusplus).
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:20]:
```suggestion
 * @brief Functions for the [Count the number of ciphers](https://www.tutorialspoint.com/count-trailing-zeros-in-factorial-of-a-number-in-cplusplus) in `n!` implementation
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+ */
+namespace numberOfCiphersInFactorialN {
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:22]:
```suggestion
namespace count_of_trailing_ciphers_in_factorial_n {
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+ /**
+ * @brief Function to count the number of trailing ciphers.
+ * @param n, Number for which `n!` ciphers are returned
+ * @return count, Number of ciphers in `n!`.
+ */
+
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:28]:
```suggestion
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    // Keep dividing n by powers of
+    // 5 and update count
+    for (uint64_t i = 5; n / i >= 1; i *= 5) {
+        count += static_cast<uint64_t>(n) / i;
+    }
+
+    return count;
+}
+}  // namespace numberOfCiphersInFactorialN
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:41]:
```suggestion
}  // namespace count_of_trailing_ciphers_in_factorial_n
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+}  // namespace bit_manipulation
+
+/**
+ * @brief Test implementations
+ * @returns void
+ */
+static void test() {
+    // Test 1
+    std::cout << "1st test ";
+    assert(bit_manipulation::numberOfCiphersInFactorialN::numberOfCiphersInFactorialN(395) == 97);
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:51]:
```suggestion
    assert(bit_manipulation::count_of_trailing_ciphers_in_factorial_n::numberOfCiphersInFactorialN(395) == 97);
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+}
+}  // namespace numberOfCiphersInFactorialN
+}  // namespace bit_manipulation
+
+/**
+ * @brief Test implementations
+ * @returns void
+ */
+static void test() {
+    // Test 1
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:49]:
```suggestion
    // 1st test
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    std::cout << "1st test ";
+    assert(bit_manipulation::numberOfCiphersInFactorialN::numberOfCiphersInFactorialN(395) == 97);
+    std::cout << "passed" << std::endl;
+
+    // Test 2
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:54]:
```suggestion
    // 2nd test
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    std::cout << "2nd test ";
+    assert(bit_manipulation::numberOfCiphersInFactorialN::numberOfCiphersInFactorialN(977) == 242);
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:56]:
```suggestion
    assert(bit_manipulation::count_of_trailing_ciphers_in_factorial_n::numberOfCiphersInFactorialN(977) == 242);
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    std::cout << "passed" << std::endl;
+
+    // Test 3
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:59]:
```suggestion
    // 3rd test
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    std::cout << "3rd test ";
+    assert(bit_manipulation::numberOfCiphersInFactorialN::numberOfCiphersInFactorialN(871) == 215);
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:61]:
```suggestion
    assert(bit_manipulation::count_of_trailing_ciphers_in_factorial_n::numberOfCiphersInFactorialN(871) == 215);
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    std::cout << "passed" << std::endl;
+
+    // Test 4
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:64]:
```suggestion
    // 4th test
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+    std::cout << "4th test ";
+    assert(bit_manipulation::numberOfCiphersInFactorialN::numberOfCiphersInFactorialN(239) == 57);
+    std::cout << "passed" << std::endl;
+
+    // Test 5
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:69]:
```suggestion
    // 5th test
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    std::cout << "passed" << std::endl;
+
+    // Test 3
+    std::cout << "3rd test ";
+    assert(bit_manipulation::numberOfCiphersInFactorialN::numberOfCiphersInFactorialN(871) == 215);
+    std::cout << "passed" << std::endl;
+
+    // Test 4
+    std::cout << "4th test ";
+    assert(bit_manipulation::numberOfCiphersInFactorialN::numberOfCiphersInFactorialN(239) == 57);
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:66]:
```suggestion
    assert(bit_manipulation::count_of_trailing_ciphers_in_factorial_n::numberOfCiphersInFactorialN(239) == 57);
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+    std::cout << "passed" << std::endl;
+
+    // Test 5
+    std::cout << "5th test ";
+    assert(bit_manipulation::numberOfCiphersInFactorialN::numberOfCiphersInFactorialN(0) == 0);
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:71]:
```suggestion
    assert(bit_manipulation::count_of_trailing_ciphers_in_factorial_n::numberOfCiphersInFactorialN(0) == 0);
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄👍
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/Swastyy/C-Plus-Plus/actions

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 18]
```diff
@@ -0,0 +1,99 @@
+/**
+ * @file
+ * @brief [Count the number of
+ ciphers](https://www.tutorialspoint.com/count-trailing-zeros-in-factorial-of-a-number-in-cplusplus)
+ in `n!` implementation
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:5]:
```suggestion
 * @brief [Count the number of 
 * ciphers](https://www.tutorialspoint.com/count-trailing-zeros-in-factorial-of-a-number-in-cplusplus) in `n!` implementation
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+ zeroes in the factorial calculated for
+ * that number. A factorial of a number N is a product of all numbers in the
+ range [1, N].
+
+ * We know that we get a trailing zero only if the number is multiple of 10 or
+ has a factor pair (2,5). In all factorials of
+ * any number greater than 5, we have many 2s more than 5s in the prime
+ factorization of that number. Dividing a
+ * number by powers of 5 will give us the count of 5s in its factors. So, the
+ number of 5s will tell us the number of trailing zeroes.
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:17]:
```suggestion
 * We know that we get a trailing zero only if the number is multiple of 10 or has a factor pair (2,5). In all factorials of
 * any number greater than 5, we have many 2s more than 5s in the prime factorization of that number. Dividing a
 * number by powers of 5 will give us the count of 5s in its factors. So, the number of 5s will tell us the number of trailing zeroes.
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+/**
+ * @namespace count_of_trailing_ciphers_in_factorial_n
+ * @brief Functions for the [Count the number of
+ * ciphers](https://www.tutorialspoint.com/count-trailing-zeros-in-factorial-of-a-number-in-cplusplus)
+ * in `n!` implementation
+ */
+namespace count_of_trailing_ciphers_in_factorial_n {
+/**
+ * @brief Function to count the number of trailing ciphers.
+ * @param n, Number for which `n!` ciphers are returned
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:38]:
```suggestion
 * @param n number for which `n!` ciphers are returned
```

File: bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+namespace bit_manipulation {
+/**
+ * @namespace count_of_trailing_ciphers_in_factorial_n
+ * @brief Functions for the [Count the number of
+ * ciphers](https://www.tutorialspoint.com/count-trailing-zeros-in-factorial-of-a-number-in-cplusplus)
+ * in `n!` implementation
+ */
+namespace count_of_trailing_ciphers_in_factorial_n {
+/**
+ * @brief Function to count the number of trailing ciphers.
```
Comment (Panquesito7) [inline bit_manipulation/count_of_trailing_ciphers_in_factorial_n.cpp:37]:
```suggestion
 * @brief Function to count the number of the trailing ciphers
```

**Review by Panquesito7 — COMMENTED** [review]
Almost there! 😄

**Review by Panquesito7 — APPROVED** [review]
LGTM. Great work, @Swastyy; thank you! 😄🎉

**Review by mishraabhinn — CHANGES_REQUESTED** [review]
Screenshot 2021-07-26 at 00 25 19
I till now don't get how the files getting deleted and changed. If you merge them @Panquesito7 this will affect repo. Github action will change files when you have made changes in that files than only.

**Comment by Panquesito7** [comment]
> GitHub action will change files when you have made changes in that files then only

It doesn't always do that. It sometimes edits other files as they are not to our standards, and the CI wants to fix it. @ayaankhan98, any ideas of what we should do? Thanks.

**Comment by Swastyy** [comment]
> Screenshot 2021-07-26 at 00 25 19
> I till now don't get how the files getting deleted and changed. If you merge them @Panquesito7 this will affect repo. Github action will change files when you have made changes in that files than only.

Now they aren't here. 😕

**Comment by mishraabhinn** [comment]
> > Screenshot 2021-07-26 at 00 25 19
> > I till now don't get how the files getting deleted and changed. If you merge them @Panquesito7 this will affect repo. Github action will change files when you have made changes in that files than only.
> 
> Now they aren't here. 😕

And, how you fixed that?

**Comment by mishraabhinn** [comment]
> > > Screenshot 2021-07-26 at 00 25 19
> > > I till now don't get how the files getting deleted and changed. If you merge them @Panquesito7 this will affect repo. Github action will change files when you have made changes in that files than only.
> > 
> > 
> > Now they aren't here. 😕
> 
> And, how you fixed that?

You removed that files from your branch na. thats why that deleted icon was coming and now you copied these same file and pasted at the exact location na?:monocle_face:

**Comment by Swastyy** [comment]
> > You removed that files from your branch na. thats why that deleted icon was coming and now you copied these same file and pasted at the exact location na?🧐

Nai nai 😂

**Comment by Swastyy** [comment]
> And, how you fixed that?
I was just trying to fix this https://stackoverflow.com/questions/41283955/github-keeps-saying-this-branch-is-x-commits-ahead-y-commits-behind
image
You did something ryt?

**Comment by mishraabhinn** [comment]
> > And, how you fixed that?
> > I was just trying to fix this https://stackoverflow.com/questions/41283955/github-keeps-saying-this-branch-is-x-commits-ahead-y-commits-behind
> > image
> > You did something ryt?

Why check_prime.cpp is there? Something majorly wrong in your way of doing commit. You are not pulling changes, i guess.

**Comment by Swastyy** [comment]
> Why check_prime.cpp is there?
Idk 😕 
> Something majorly wrong in your way of doing commit. You are not pulling changes, i guess.
I usually click on fetch and merge button since I use gitpod here is it required to pull from gitpod's command line too?

**Comment by mishraabhinn** [comment]
> > Why check_prime.cpp is there?
> > Idk 😕
> > Something majorly wrong in your way of doing commit. You are not pulling changes, i guess.
> > I usually click on fetch and merge button since I use gitpod here is it required to pull from gitpod's command line too?

Lets approve this up.  As changes are made by github action bot and they seems to be correct and valid. Nice work @Swastyy :partying_face: . You can try using vs / terminal

**Review by mishraabhinn — APPROVED** [review]

**Review by Panquesito7 — APPROVED** [review]

**Pull request merged**